#  IPL Matches — Pandas Data Analysis

## About This Notebook

This notebook demonstrates **core Pandas concepts** using the IPL (Indian Premier League) matches dataset (`matches.csv`). Each code cell is preceded by a Markdown explanation describing what the code does, why the technique is used in real-world data analysis, and what it reveals about the IPL dataset.

**Dataset:** IPL matches from 2008 to 2017 — 636 matches, 18 columns  
**Topics covered:** Data loading, inspection, selection, filtering, sorting, deduplication, grouping, and missing value handling.

---

##  Step 1: Importing the Pandas Library

**What is happening here?**
This line imports the Pandas library and gives it the alias `pd`. This alias is a universal convention in the Python data science community — it saves us from typing `pandas` every time we use it.

**Real-world use:**
Every data analysis project in Python starts with this line. Pandas is the backbone of data manipulation and analysis, used by data analysts, data scientists, and researchers to work with structured data (tables, spreadsheets, databases).

**Purpose in this notebook:**
Without this import, none of the Pandas methods and functions used throughout this notebook would work. It is the foundation on which everything else is built.

In [ ]:
import pandas as pd

##  Step 2: Loading the Dataset

**What is happening here?**
`pd.read_csv()` reads a CSV (Comma Separated Values) file from disk and loads it into a Pandas **DataFrame** — a two-dimensional table structure with rows and columns, similar to an Excel spreadsheet. The result is stored in a variable called `data`.

**Real-world use:**
Loading CSV files is one of the most common tasks in data analysis. CSV is a universal format used for storing structured data — sales records, survey results, sports statistics, financial data, and more. `read_csv()` also supports parameters for handling delimiters, encodings, date parsing, and more.

**Purpose in this notebook:**
This loads the IPL (Indian Premier League) matches dataset (`matches.csv`) into memory. All subsequent analysis in this notebook operates on this `data` object.

In [ ]:
data = pd.read_csv('matches.csv')

##  Step 3: Displaying the Entire DataFrame

**What is happening here?**
Simply typing `data` in a Jupyter Notebook cell displays the full DataFrame in a formatted table. Jupyter automatically shows the first and last 5 rows when the table is large, along with a summary of the total number of rows and columns.

**Real-world use:**
Getting a quick visual overview of your data is always the first thing analysts do after loading a dataset. It helps you immediately understand the structure, column names, data types, and get a feel for what the data looks like.

**Purpose in this notebook:**
This reveals that the IPL dataset has **636 rows and 18 columns**, covering match details such as teams, city, date, toss information, winner, win margin, players of the match, venues, and umpires.

In [ ]:
data

,id,season,city,date,team1,team2,toss_winner,toss_decision,result,dl_applied,winner,win_by_runs,win_by_wickets,player_of_match,venue,umpire1,umpire2,umpire3
0,1,2017,Hyderabad,2017-04-05,Sunrisers Hyderabad,Royal Challengers Bangalore,Royal Challengers Bangalore,field,normal,0,Sunrisers Hyderabad,35,0,Yuvraj Singh,"Rajiv Gandhi International Stadium, Uppal",AY Dandekar,NJ Llong,NaN
1,2,2017,Pune,2017-04-06,Mumbai Indians,Rising Pune Supergiant,Rising Pune Supergiant,field,normal,0,Rising Pune Supergiant,0,7,SPD Smith,Maharashtra Cricket Association Stadium,A Nand Kishore,S Ravi,NaN
2,3,2017,Rajkot,2017-04-07,Gujarat Lions,Kolkata Knight Riders,Kolkata Knight Riders,field,normal,0,Kolkata Knight Riders,0,10,CA Lynn,Saurashtra Cricket Association Stadium,Nitin Menon,CK Nandan,NaN
3,4,2017,Indore,2017-04-08,Rising Pune Supergiant,Kings XI Punjab,Kings XI Punjab,field,normal,0,Kings XI Punjab,0,6,GJ Maxwell,Holkar Cricket Stadium,AK Chaudhary,C Shamshuddin,NaN
4,5,2017,Bangalore,2017-04-08,Royal Challengers Bangalore,Delhi Daredevils,Royal Challengers Bangalore,bat,normal,0,Royal Challengers Bangalore,15,0,KM Jadhav,M Chinnaswamy Stadium,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
631,632,2016,Raipur,2016-05-22,Delhi Daredevils,Royal Challengers Bangalore,Royal Challengers Bangalore,field,normal,0,Royal Challengers Bangalore,0,6,V Kohli,Shaheed Veer Narayan Singh International Stadium,A Nand Kishore,BNJ Oxenford,NaN
632,633,2016,Bangalore,2016-05-24,Gujarat Lions,Royal Challengers Bangalore,Royal Challengers Bangalore,field,normal,0,Royal Challengers Bangalore,0,4,AB de Villiers,M Chinnaswamy Stadium,AK Chaudhary,HDPK Dharmasena,NaN
633,634,2016,Delhi,2016-05-25,Sunrisers Hyderabad,Kolkata Knight Riders,Kolkata Knight Riders,field,normal,0,Sunrisers Hyderabad,22,0,MC Henriques,Feroz Shah Kotla,M Erasmus,C Shamshuddin,NaN
634,635,2016,Delhi,2016-05-27,Gujarat Lions,Sunrisers Hyderabad,Sunrisers Hyderabad,field,normal,0,Sunrisers Hyderabad,0,4,DA Warner,Feroz Shah Kotla,M Erasmus,CK Nandan,NaN


## Step 4: Previewing the First 5 Rows with `head()`

**What is happening here?**
`data.head()` returns only the **first 5 rows** of the DataFrame (you can pass a number like `head(10)` to see more). This is a much lighter operation than displaying the full dataset.

**Real-world use:**
`head()` is used constantly in data analysis to do a quick sanity check — confirming that data was loaded correctly, column names make sense, and that the values look reasonable. It is faster and cleaner than printing the whole dataset.

**Purpose in this notebook:**
After loading the IPL data, `head()` lets us confirm the first few matches, their dates, teams, and outcomes, without scrolling through all 636 rows.

In [ ]:
data.head()

,id,season,city,date,team1,team2,toss_winner,toss_decision,result,dl_applied,winner,win_by_runs,win_by_wickets,player_of_match,venue,umpire1,umpire2,umpire3
0,1,2017,Hyderabad,2017-04-05,Sunrisers Hyderabad,Royal Challengers Bangalore,Royal Challengers Bangalore,field,normal,0,Sunrisers Hyderabad,35,0,Yuvraj Singh,"Rajiv Gandhi International Stadium, Uppal",AY Dandekar,NJ Llong,NaN
1,2,2017,Pune,2017-04-06,Mumbai Indians,Rising Pune Supergiant,Rising Pune Supergiant,field,normal,0,Rising Pune Supergiant,0,7,SPD Smith,Maharashtra Cricket Association Stadium,A Nand Kishore,S Ravi,NaN
2,3,2017,Rajkot,2017-04-07,Gujarat Lions,Kolkata Knight Riders,Kolkata Knight Riders,field,normal,0,Kolkata Knight Riders,0,10,CA Lynn,Saurashtra Cricket Association Stadium,Nitin Menon,CK Nandan,NaN
3,4,2017,Indore,2017-04-08,Rising Pune Supergiant,Kings XI Punjab,Kings XI Punjab,field,normal,0,Kings XI Punjab,0,6,GJ Maxwell,Holkar Cricket Stadium,AK Chaudhary,C Shamshuddin,NaN
4,5,2017,Bangalore,2017-04-08,Royal Challengers Bangalore,Delhi Daredevils,Royal Challengers Bangalore,bat,normal,0,Royal Challengers Bangalore,15,0,KM Jadhav,M Chinnaswamy Stadium,NaN,NaN,NaN


## Step 5: Previewing the Last 5 Rows with `tail()`

**What is happening here?**
`data.tail()` returns the **last 5 rows** of the DataFrame. Similar to `head()`, you can pass a number to see more rows.

**Real-world use:**
`tail()` is useful for checking if data was appended correctly, verifying the end of time-series data, or confirming that no empty rows were added at the bottom of a file during export.

**Purpose in this notebook:**
This confirms the last few IPL matches in the dataset (from the 2016 season), helping verify that the data ends where we expect it to.

In [ ]:
data.tail()

,id,season,city,date,team1,team2,toss_winner,toss_decision,result,dl_applied,winner,win_by_runs,win_by_wickets,player_of_match,venue,umpire1,umpire2,umpire3
631,632,2016,Raipur,2016-05-22,Delhi Daredevils,Royal Challengers Bangalore,Royal Challengers Bangalore,field,normal,0,Royal Challengers Bangalore,0,6,V Kohli,Shaheed Veer Narayan Singh International Stadium,A Nand Kishore,BNJ Oxenford,NaN
632,633,2016,Bangalore,2016-05-24,Gujarat Lions,Royal Challengers Bangalore,Royal Challengers Bangalore,field,normal,0,Royal Challengers Bangalore,0,4,AB de Villiers,M Chinnaswamy Stadium,AK Chaudhary,HDPK Dharmasena,NaN
633,634,2016,Delhi,2016-05-25,Sunrisers Hyderabad,Kolkata Knight Riders,Kolkata Knight Riders,field,normal,0,Sunrisers Hyderabad,22,0,MC Henriques,Feroz Shah Kotla,M Erasmus,C Shamshuddin,NaN
634,635,2016,Delhi,2016-05-27,Gujarat Lions,Sunrisers Hyderabad,Sunrisers Hyderabad,field,normal,0,Sunrisers Hyderabad,0,4,DA Warner,Feroz Shah Kotla,M Erasmus,CK Nandan,NaN
635,636,2016,Bangalore,2016-05-29,Sunrisers Hyderabad,Royal Challengers Bangalore,Sunrisers Hyderabad,bat,normal,0,Sunrisers Hyderabad,8,0,BCJ Cutting,M Chinnaswamy Stadium,HDPK Dharmasena,BNJ Oxenford,NaN


##  Step 6: Checking the Shape of the DataFrame

**What is happening here?**
`data.shape` returns a tuple `(rows, columns)` — in this case `(636, 18)`. It is a quick way to know how large your dataset is.

**Real-world use:**
Checking the shape of a DataFrame is a standard step after loading data. Before any analysis, it is important to know how many observations (rows) and features (columns) you are working with. This also helps detect if data was accidentally lost or duplicated during loading or processing.

**Purpose in this notebook:**
Confirms the IPL dataset has 636 matches and 18 attributes per match.

In [ ]:
data.shape

(636, 18)

##  Step 7: Getting Structural Information with `info()`

**What is happening here?**
`data.info()` prints a concise summary of the DataFrame, including:
- The total number of rows and columns
- Each column's name and data type (`dtype`)
- The number of **non-null** (non-missing) values per column

**Real-world use:**
`info()` is one of the most important first steps in Exploratory Data Analysis (EDA). It immediately tells you which columns have missing values and what data types are being used. For example, it can reveal that a numeric column was accidentally stored as text (`object`), or that a date column needs to be converted.

**Purpose in this notebook:**
We can see that `umpire3` has 0 non-null values (entirely empty), `winner` has 633 (3 missing), `city` has 629 (7 missing), and `player_of_match` also has some missing values. This shapes our data cleaning decisions later.

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 18 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id               636 non-null    int64  
 1   season           636 non-null    int64  
 2   city             629 non-null    object 
 3   date             636 non-null    object 
 4   team1            636 non-null    object 
 5   team2            636 non-null    object 
 6   toss_winner      636 non-null    object 
 7   toss_decision    636 non-null    object 
 8   result           636 non-null    object 
 9   dl_applied       636 non-null    int64  
 10  winner           633 non-null    object 
 11  win_by_runs      636 non-null    int64  
 12  win_by_wickets   636 non-null    int64  
 13  player_of_match  633 non-null    object 
 14  venue            636 non-null    object 
 15  umpire1          635 non-null    object 
 16  umpire2          635 non-null    object 
 17  umpire3         

##  Step 8: Statistical Summary with `describe()`

**What is happening here?**
`data.describe()` generates descriptive statistics for all **numeric columns** in the DataFrame. It shows count, mean, standard deviation, minimum, maximum, and the 25th/50th/75th percentiles (quartiles).

**Real-world use:**
`describe()` is a cornerstone of EDA. It helps you immediately spot potential issues such as outliers (very high max values), skewed distributions (large gap between mean and median), or zero-variance columns. In financial, medical, and sports analytics, this step is essential for understanding data ranges and distributions.

**Purpose in this notebook:**
We can see that `win_by_runs` ranges from 0 to 146, with a mean of ~13.7. `win_by_wickets` ranges from 0 to 10. Notably, `umpire3` shows a count of 0, confirming it's an entirely empty column.

In [ ]:
data.describe()

,id,season,dl_applied,win_by_runs,win_by_wickets,umpire3
count,636.000000,636.000000,636.000000,636.000000,636.000000,0.0
mean,318.500000,2012.490566,0.025157,13.682390,3.372642,NaN
std,183.741666,2.773026,0.156726,23.908877,3.420338,NaN
min,1.000000,2008.000000,0.000000,0.000000,0.000000,NaN
25%,159.750000,2010.000000,0.000000,0.000000,0.000000,NaN
50%,318.500000,2012.000000,0.000000,0.000000,4.000000,NaN
75%,477.250000,2015.000000,0.000000,20.000000,7.000000,NaN
max,636.000000,2017.000000,1.000000,146.000000,10.000000,NaN


## Step 9: Selecting a Single Column

**What is happening here?**
`data['team1']` selects and returns a single column from the DataFrame as a **Pandas Series** — a one-dimensional array with an index. Here, it returns all 636 team names for the first team in each match.

**Real-world use:**
Selecting specific columns is a fundamental operation. In real projects, you rarely need all columns — you pick only the ones relevant to your current analysis. For example, in a sales dataset you might select only the `revenue` column to calculate total sales.

**Purpose in this notebook:**
This extracts the `team1` column to inspect which teams appear as the "home" or first-listed team across all matches.

In [ ]:
data['team1']

,team1
0,Sunrisers Hyderabad
1,Mumbai Indians
2,Gujarat Lions
3,Rising Pune Supergiant
4,Royal Challengers Bangalore
...,...
631,Delhi Daredevils
632,Gujarat Lions
633,Sunrisers Hyderabad
634,Gujarat Lions


##  Step 10: Selecting Multiple Columns

**What is happening here?**
By passing a **list of column names** inside `data[...]`, we select multiple columns at once and get back a smaller DataFrame (a subset of the original). Here, we extract `team1`, `team2`, and `winner` together.

**Real-world use:**
Selecting multiple columns is used to focus analysis on relevant variables, reduce memory usage, and create clean views of data for reporting. In machine learning, this is also used to separate features from target labels.

**Purpose in this notebook:**
This creates a focused view showing just the two competing teams and the match winner — perfect for analyzing head-to-head records or win patterns.

In [ ]:
data[['team1','team2','winner']]

,team1,team2,winner
0,Sunrisers Hyderabad,Royal Challengers Bangalore,Sunrisers Hyderabad
1,Mumbai Indians,Rising Pune Supergiant,Rising Pune Supergiant
2,Gujarat Lions,Kolkata Knight Riders,Kolkata Knight Riders
3,Rising Pune Supergiant,Kings XI Punjab,Kings XI Punjab
4,Royal Challengers Bangalore,Delhi Daredevils,Royal Challengers Bangalore
...,...,...,...
631,Delhi Daredevils,Royal Challengers Bangalore,Royal Challengers Bangalore
632,Gujarat Lions,Royal Challengers Bangalore,Royal Challengers Bangalore
633,Sunrisers Hyderabad,Kolkata Knight Riders,Sunrisers Hyderabad
634,Gujarat Lions,Sunrisers Hyderabad,Sunrisers Hyderabad


##  Step 11: Accessing a Row by Position using `iloc`

**What is happening here?**
`data.iloc[1]` uses **integer-location based indexing** to retrieve a single row by its position (index 1 = the second row). It returns the row as a Series with column names as the index.

**Real-world use:**
`iloc` is used when you want to access data by its numerical position in the DataFrame, regardless of what the index label is. This is particularly useful after filtering or reindexing operations, where the original row numbers may not be sequential.

**Purpose in this notebook:**
This retrieves all details of the second match in the dataset (Mumbai Indians vs Rising Pune Supergiant, played on 2017-04-06).

In [ ]:
data.iloc[1]

,1
id,2
season,2017
city,Pune
date,2017-04-06
team1,Mumbai Indians
team2,Rising Pune Supergiant
toss_winner,Rising Pune Supergiant
toss_decision,field
result,normal
dl_applied,0


##  Step 12: Selecting Specific Columns by Position using `iloc`

**What is happening here?**
`data.iloc[:, [4,5,10]]` selects all rows (`:`) but only the columns at positions 4, 5, and 10 (team1, team2, winner). This uses pure integer-based indexing for both rows and columns.

**Real-world use:**
`iloc` with column indices is helpful when you know the position of a column but not its exact name, or when working with large datasets where columns are numbered. It is also useful in loops or automated pipelines where column positions are calculated programmatically.

**Purpose in this notebook:**
This achieves the same result as selecting by name (`data[['team1','team2','winner']]`), demonstrating the positional alternative approach.

In [ ]:
data.iloc[:,[4,5,10]]

,team1,team2,winner
0,Sunrisers Hyderabad,Royal Challengers Bangalore,Sunrisers Hyderabad
1,Mumbai Indians,Rising Pune Supergiant,Rising Pune Supergiant
2,Gujarat Lions,Kolkata Knight Riders,Kolkata Knight Riders
3,Rising Pune Supergiant,Kings XI Punjab,Kings XI Punjab
4,Royal Challengers Bangalore,Delhi Daredevils,Royal Challengers Bangalore
...,...,...,...
631,Delhi Daredevils,Royal Challengers Bangalore,Royal Challengers Bangalore
632,Gujarat Lions,Royal Challengers Bangalore,Royal Challengers Bangalore
633,Sunrisers Hyderabad,Kolkata Knight Riders,Sunrisers Hyderabad
634,Gujarat Lions,Sunrisers Hyderabad,Sunrisers Hyderabad


##  Step 13: Boolean Filtering (Masking) — Single Condition

**What is happening here?**
A **boolean mask** is created by comparing the `city` column to `'Hyderabad'`. This produces a Series of `True`/`False` values. Applying this mask to the DataFrame (`data[mask]`) returns only the rows where the condition is `True`. `.shape` then tells us how many rows match.

**Real-world use:**
Boolean filtering (also called masking) is one of the most powerful and frequently used Pandas techniques. It is the equivalent of a `WHERE` clause in SQL. In real projects, you use it to filter customers by region, transactions by date range, or patients by diagnosis.

**Purpose in this notebook:**
This counts the number of IPL matches played in Hyderabad — the result is 49 matches.

In [ ]:
mask=data['city']=='Hyderabad'
data[mask].shape

(49, 18)

##  Step 14: Wrapping Logic in a Reusable Function

**What is happening here?**
A Python function `match_count(city)` is defined that takes a city name as input, creates a boolean mask for that city, and returns the count of matches played there. This encapsulates the filtering logic from the previous step.

**Real-world use:**
Wrapping repetitive data analysis logic into functions is a best practice in software engineering and data science. It avoids code duplication, makes code more readable, easier to test, and allows you to quickly answer the same question for different inputs (e.g., match counts for Mumbai, Chennai, etc.).

**Purpose in this notebook:**
This function makes it easy to query the match count for any city, demonstrating the value of reusability in data analysis workflows.

In [ ]:
def match_count(city):
  mask=data['city']=='Hyderabad'
  return data[mask].shape[0]

## Step 15: Calling the Function

**What is happening here?**
The `match_count` function is called with `'Hyderabad'` as the argument. It applies the filter logic defined in the previous cell and returns `49`.

**Real-world use:**
Calling functions with different parameters enables quick comparative analysis — e.g., comparing match counts across cities without rewriting the filtering logic each time.

**Purpose in this notebook:**
Confirms the function works correctly and demonstrates how abstracted logic can produce clean, readable outputs.

In [ ]:
match_count('Hyderabad')

49

##  Step 16: Boolean Filtering with Multiple Conditions (`&` operator)

**What is happening here?**
Two separate boolean masks are created — one for city (`Hyderabad`) and one for date (after 2010-01-01). The `&` operator combines them using a logical AND, so only rows satisfying **both** conditions are returned.

**Real-world use:**
Combining multiple filter conditions is extremely common. In SQL this is `WHERE city = 'X' AND date > 'Y'`. In Pandas you use `&` for AND and `|` for OR. This is used in e-commerce (filter by category AND price range), finance (filter by asset class AND date), and healthcare (filter by diagnosis AND age group).

**Purpose in this notebook:**
Filters IPL matches played in Hyderabad after the year 2010, which returns 42 matches from the 2011–2017 seasons.

In [ ]:
mask1=data['city']=='Hyderabad'
mask2=data['date']>'2010-01-01'
data[mask1 & mask2]

,id,season,city,date,team1,team2,toss_winner,toss_decision,result,dl_applied,winner,win_by_runs,win_by_wickets,player_of_match,venue,umpire1,umpire2,umpire3
0,1,2017,Hyderabad,2017-04-05,Sunrisers Hyderabad,Royal Challengers Bangalore,Royal Challengers Bangalore,field,normal,0,Sunrisers Hyderabad,35,0,Yuvraj Singh,"Rajiv Gandhi International Stadium, Uppal",AY Dandekar,NJ Llong,NaN
5,6,2017,Hyderabad,2017-04-09,Gujarat Lions,Sunrisers Hyderabad,Sunrisers Hyderabad,field,normal,0,Sunrisers Hyderabad,0,9,Rashid Khan,"Rajiv Gandhi International Stadium, Uppal",A Deshmukh,NJ Llong,NaN
18,19,2017,Hyderabad,2017-04-17,Sunrisers Hyderabad,Kings XI Punjab,Kings XI Punjab,field,normal,0,Sunrisers Hyderabad,5,0,B Kumar,"Rajiv Gandhi International Stadium, Uppal",AY Dandekar,A Deshmukh,NaN
20,21,2017,Hyderabad,2017-04-19,Sunrisers Hyderabad,Delhi Daredevils,Sunrisers Hyderabad,bat,normal,0,Sunrisers Hyderabad,15,0,KS Williamson,"Rajiv Gandhi International Stadium, Uppal",CB Gaffaney,NJ Llong,NaN
35,36,2017,Hyderabad,2017-04-30,Sunrisers Hyderabad,Kolkata Knight Riders,Kolkata Knight Riders,field,normal,0,Sunrisers Hyderabad,48,0,DA Warner,"Rajiv Gandhi International Stadium, Uppal",AY Dandekar,S Ravi,NaN
42,43,2017,Hyderabad,2017-05-06,Rising Pune Supergiant,Sunrisers Hyderabad,Sunrisers Hyderabad,field,normal,0,Rising Pune Supergiant,12,0,JD Unadkat,"Rajiv Gandhi International Stadium, Uppal",KN Ananthapadmanabhan,AK Chaudhary,NaN
46,47,2017,Hyderabad,2017-05-08,Mumbai Indians,Sunrisers Hyderabad,Mumbai Indians,bat,normal,0,Sunrisers Hyderabad,0,7,S Dhawan,"Rajiv Gandhi International Stadium, Uppal",KN Ananthapadmanabhan,M Erasmus,NaN
58,59,2017,Hyderabad,2017-05-21,Mumbai Indians,Rising Pune Supergiant,Mumbai Indians,bat,normal,0,Mumbai Indians,1,0,KH Pandya,"Rajiv Gandhi International Stadium, Uppal",NJ Llong,S Ravi,NaN
235,236,2011,Hyderabad,2011-04-09,Deccan Chargers,Rajasthan Royals,Rajasthan Royals,field,normal,0,Rajasthan Royals,0,8,SK Trivedi,"Rajiv Gandhi International Stadium, Uppal",RE Koertzen,SK Tarapore,NaN
244,245,2011,Hyderabad,2011-04-14,Deccan Chargers,Royal Challengers Bangalore,Royal Challengers Bangalore,field,normal,0,Deccan Chargers,33,0,DW Steyn,"Rajiv Gandhi International Stadium, Uppal",RE Koertzen,S Ravi,NaN


##  Step 17: Counting Frequencies with `value_counts()`

**What is happening here?**
`value_counts()` returns a Series showing the **count of each unique value** in the `winner` column, sorted from most to least frequent by default.

**Real-world use:**
`value_counts()` is one of the most used Pandas methods. It is the equivalent of `GROUP BY column COUNT(*)` in SQL. Used in market analysis (most popular products), sports analytics (team performance), social media analysis (most used hashtags), and HR analytics (distribution of job roles).

**Purpose in this notebook:**
This instantly reveals that **Mumbai Indians** has won the most matches (92), followed by Chennai Super Kings (79) and Kolkata Knight Riders (77).

In [ ]:
data['winner'].value_counts()

,count
winner,
Mumbai Indians,92
Chennai Super Kings,79
Kolkata Knight Riders,77
Royal Challengers Bangalore,73
Kings XI Punjab,70
Rajasthan Royals,63
Delhi Daredevils,62
Sunrisers Hyderabad,42
Deccan Chargers,29


##  Step 18: Storing Results in a Variable

**What is happening here?**
The result of `value_counts()` (a Series) is stored in a variable called `myseries`. This is a common pattern — compute an aggregation once and reuse it in multiple subsequent operations without re-running the computation.

**Real-world use:**
Storing intermediate results in variables is a fundamental programming practice. It avoids redundant computation, especially important with large datasets or complex aggregations.

**Purpose in this notebook:**
`myseries` is used in the next several cells to demonstrate Series-specific operations like `.index`, `.values`, and label-based indexing.

In [ ]:
myseries= data['winner'].value_counts()

##  Step 19: Accessing the Index of a Series

**What is happening here?**
`.index` is an attribute of a Pandas Series (or DataFrame) that returns the **labels/names of the rows**. For `myseries`, the index is the list of team names.

**Real-world use:**
Understanding and working with the index is fundamental in Pandas. The index enables fast lookups, label-based alignment when merging datasets, and time-series operations. In real projects, you might reset, rename, or set a custom index to make data access more efficient.

**Purpose in this notebook:**
Shows the list of all 14 IPL teams that have won at least one match in the dataset.

In [ ]:
myseries.index

Index(['Mumbai Indians', 'Chennai Super Kings', 'Kolkata Knight Riders',
       'Royal Challengers Bangalore', 'Kings XI Punjab', 'Rajasthan Royals',
       'Delhi Daredevils', 'Sunrisers Hyderabad', 'Deccan Chargers',
       'Gujarat Lions', 'Pune Warriors', 'Rising Pune Supergiant',
       'Kochi Tuskers Kerala', 'Rising Pune Supergiants'],
      dtype='object', name='winner')

##  Step 20: Accessing the Values of a Series

**What is happening here?**
`.values` returns the underlying data of a Pandas Series as a **NumPy array** — stripping away the index labels and giving you just the raw numbers.

**Real-world use:**
You use `.values` when you need to pass data to libraries that don't understand Pandas (e.g., NumPy, scikit-learn, matplotlib) or when you want to perform fast numerical operations without the overhead of index alignment.

**Purpose in this notebook:**
Extracts the raw win counts for all teams as a NumPy array: `[92, 79, 77, 73, ...]`.

In [ ]:
myseries.values

array([92, 79, 77, 73, 70, 63, 62, 42, 29, 13, 12, 10,  6,  5])

## Step 21: Label-Based Access in a Series

**What is happening here?**
When a Series has string labels as its index (like team names), you can access a specific value using its label inside square brackets — just like a dictionary lookup. Here, it retrieves the total number of wins for Kings XI Punjab (70).

**Real-world use:**
Label-based lookups are used when you need to retrieve specific aggregated values by name — e.g., the total revenue of a specific product, or the vote count for a specific candidate.

**Purpose in this notebook:**
Quickly retrieves the win total for a specific team without writing a filter.

In [ ]:
myseries['Kings XI Punjab']

np.int64(70)

##  Step 22: Sorting Values in Ascending Order

**What is happening here?**
`sort_values()` reorders a Series by its values in **ascending order** (smallest to largest) by default. The index labels (team names) move along with their corresponding values.

**Real-world use:**
Sorting is fundamental in any ranking or leaderboard scenario. Ascending sort is useful when identifying the lowest-performing entities — lowest sales regions, least popular products, or teams with fewest wins.

**Purpose in this notebook:**
Produces a ranking of IPL teams from fewest wins to most wins.

In [ ]:
myseries.sort_values()

,count
winner,
Rising Pune Supergiants,5
Kochi Tuskers Kerala,6
Rising Pune Supergiant,10
Pune Warriors,12
Gujarat Lions,13
Deccan Chargers,29
Sunrisers Hyderabad,42
Delhi Daredevils,62
Rajasthan Royals,63


##  Step 23: Sorting Values in Descending Order

**What is happening here?**
Passing `ascending=False` to `sort_values()` reverses the sort order, placing the highest values at the top.

**Real-world use:**
Descending sort is used to build top-N lists or leaderboards — top 10 customers by revenue, highest-rated products, most active users, or in this case, the most successful IPL teams.

**Purpose in this notebook:**
Confirms the ranking of IPL teams by wins with Mumbai Indians at the top (92 wins), making this an easy way to identify the most successful franchise.

In [ ]:
myseries.sort_values(ascending=False)

,count
winner,
Mumbai Indians,92
Chennai Super Kings,79
Kolkata Knight Riders,77
Royal Challengers Bangalore,73
Kings XI Punjab,70
Rajasthan Royals,63
Delhi Daredevils,62
Sunrisers Hyderabad,42
Deccan Chargers,29


##  Step 24: Sorting a DataFrame by a Column

**What is happening here?**
`data.sort_values('city')` sorts the **entire DataFrame** alphabetically by the values in the `city` column (ascending by default). All other columns remain aligned with their corresponding rows.

**Real-world use:**
Sorting DataFrames by a column is used constantly — sort transactions by date, customers by name, products by price. It is the Pandas equivalent of `ORDER BY` in SQL. It is also useful before grouping operations or for visual presentation of reports.

**Purpose in this notebook:**
Reorders all 636 IPL matches alphabetically by city name, making it easy to visually group matches that occurred in the same city.

In [ ]:
data.sort_values('city')

,id,season,city,date,team1,team2,toss_winner,toss_decision,result,dl_applied,winner,win_by_runs,win_by_wickets,player_of_match,venue,umpire1,umpire2,umpire3
460,461,2014,Abu Dhabi,2014-04-18,Sunrisers Hyderabad,Rajasthan Royals,Rajasthan Royals,field,normal,0,Rajasthan Royals,0,4,AM Rahane,Sheikh Zayed Stadium,BF Bowden,RK Illingworth,NaN
471,472,2014,Abu Dhabi,2014-04-26,Kings XI Punjab,Kolkata Knight Riders,Kolkata Knight Riders,field,normal,0,Kings XI Punjab,23,0,Sandeep Sharma,Sheikh Zayed Stadium,HDPK Dharmasena,RK Illingworth,NaN
470,471,2014,Abu Dhabi,2014-04-26,Royal Challengers Bangalore,Rajasthan Royals,Rajasthan Royals,field,normal,0,Rajasthan Royals,0,6,PV Tambe,Sheikh Zayed Stadium,HDPK Dharmasena,C Shamshuddin,NaN
464,465,2014,Abu Dhabi,2014-04-21,Chennai Super Kings,Delhi Daredevils,Chennai Super Kings,bat,normal,0,Chennai Super Kings,93,0,SK Raina,Sheikh Zayed Stadium,RK Illingworth,C Shamshuddin,NaN
459,460,2014,Abu Dhabi,2014-04-18,Chennai Super Kings,Kings XI Punjab,Chennai Super Kings,bat,normal,0,Kings XI Punjab,0,6,GJ Maxwell,Sheikh Zayed Stadium,RK Illingworth,C Shamshuddin,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
466,467,2014,NaN,2014-04-23,Chennai Super Kings,Rajasthan Royals,Rajasthan Royals,field,normal,0,Chennai Super Kings,7,0,RA Jadeja,Dubai International Cricket Stadium,HDPK Dharmasena,RK Illingworth,NaN
468,469,2014,NaN,2014-04-25,Sunrisers Hyderabad,Delhi Daredevils,Sunrisers Hyderabad,bat,normal,0,Sunrisers Hyderabad,4,0,AJ Finch,Dubai International Cricket Stadium,M Erasmus,S Ravi,NaN
469,470,2014,NaN,2014-04-25,Mumbai Indians,Chennai Super Kings,Mumbai Indians,bat,normal,0,Chennai Super Kings,0,7,MM Sharma,Dubai International Cricket Stadium,BF Bowden,M Erasmus,NaN
474,475,2014,NaN,2014-04-28,Royal Challengers Bangalore,Kings XI Punjab,Kings XI Punjab,field,normal,0,Kings XI Punjab,0,5,Sandeep Sharma,Dubai International Cricket Stadium,BF Bowden,S Ravi,NaN


## Step 25: Sorting a DataFrame in Descending Order

**What is happening here?**
Adding `ascending=False` to `sort_values()` on a DataFrame sorts in reverse order — here, from cities at the end of the alphabet (like Visakhapatnam) to cities at the beginning (Abu Dhabi).

**Real-world use:**
Descending sort on categorical data is useful for quick inspection of the "last" entries in a sorted dataset, or when you want cities/names in Z-to-A order for certain display purposes.

**Purpose in this notebook:**
Demonstrates the `ascending` parameter control on DataFrames.

In [ ]:
data.sort_values('city',ascending=False)

,id,season,city,date,team1,team2,toss_winner,toss_decision,result,dl_applied,winner,win_by_runs,win_by_wickets,player_of_match,venue,umpire1,umpire2,umpire3
312,313,2012,Visakhapatnam,2012-04-07,Chennai Super Kings,Deccan Chargers,Deccan Chargers,field,normal,0,Chennai Super Kings,74,0,RA Jadeja,Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket St...,JD Cloete,HDPK Dharmasena,NaN
622,623,2016,Visakhapatnam,2016-05-15,Mumbai Indians,Delhi Daredevils,Delhi Daredevils,field,normal,0,Mumbai Indians,80,0,KH Pandya,Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket St...,Nitin Menon,CK Nandan,NaN
615,616,2016,Visakhapatnam,2016-05-10,Sunrisers Hyderabad,Rising Pune Supergiants,Sunrisers Hyderabad,bat,normal,0,Sunrisers Hyderabad,4,0,A Zampa,Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket St...,CB Gaffaney,VK Sharma,NaN
612,613,2016,Visakhapatnam,2016-05-08,Sunrisers Hyderabad,Mumbai Indians,Mumbai Indians,field,normal,0,Sunrisers Hyderabad,85,0,A Nehra,Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket St...,S Ravi,C Shamshuddin,NaN
624,625,2016,Visakhapatnam,2016-05-17,Delhi Daredevils,Rising Pune Supergiants,Rising Pune Supergiants,field,normal,1,Rising Pune Supergiants,19,0,AB Dinda,Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket St...,Nitin Menon,C Shamshuddin,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
466,467,2014,NaN,2014-04-23,Chennai Super Kings,Rajasthan Royals,Rajasthan Royals,field,normal,0,Chennai Super Kings,7,0,RA Jadeja,Dubai International Cricket Stadium,HDPK Dharmasena,RK Illingworth,NaN
468,469,2014,NaN,2014-04-25,Sunrisers Hyderabad,Delhi Daredevils,Sunrisers Hyderabad,bat,normal,0,Sunrisers Hyderabad,4,0,AJ Finch,Dubai International Cricket Stadium,M Erasmus,S Ravi,NaN
469,470,2014,NaN,2014-04-25,Mumbai Indians,Chennai Super Kings,Mumbai Indians,bat,normal,0,Chennai Super Kings,0,7,MM Sharma,Dubai International Cricket Stadium,BF Bowden,M Erasmus,NaN
474,475,2014,NaN,2014-04-28,Royal Challengers Bangalore,Kings XI Punjab,Kings XI Punjab,field,normal,0,Kings XI Punjab,0,5,Sandeep Sharma,Dubai International Cricket Stadium,BF Bowden,S Ravi,NaN


## Step 26: Sorting by Multiple Columns

**What is happening here?**
Passing a **list of column names** to `sort_values()` sorts by the first column, then uses the second column to break ties within the same first column value. Here, matches are first sorted by city, then within each city sorted alphabetically by winner.

**Real-world use:**
Multi-column sorting is used in reports and analyses — sort employees by department, then by salary within each department. In SQL this is `ORDER BY city, winner`.

**Purpose in this notebook:**
Groups matches by city and within each city orders them by the winning team name — useful for quickly auditing which teams won in each city.

In [ ]:
data.sort_values(['city','winner'])

,id,season,city,date,team1,team2,toss_winner,toss_decision,result,dl_applied,winner,win_by_runs,win_by_wickets,player_of_match,venue,umpire1,umpire2,umpire3
464,465,2014,Abu Dhabi,2014-04-21,Chennai Super Kings,Delhi Daredevils,Chennai Super Kings,bat,normal,0,Chennai Super Kings,93,0,SK Raina,Sheikh Zayed Stadium,RK Illingworth,C Shamshuddin,NaN
459,460,2014,Abu Dhabi,2014-04-18,Chennai Super Kings,Kings XI Punjab,Chennai Super Kings,bat,normal,0,Kings XI Punjab,0,6,GJ Maxwell,Sheikh Zayed Stadium,RK Illingworth,C Shamshuddin,NaN
471,472,2014,Abu Dhabi,2014-04-26,Kings XI Punjab,Kolkata Knight Riders,Kolkata Knight Riders,field,normal,0,Kings XI Punjab,23,0,Sandeep Sharma,Sheikh Zayed Stadium,HDPK Dharmasena,RK Illingworth,NaN
457,458,2014,Abu Dhabi,2014-04-16,Kolkata Knight Riders,Mumbai Indians,Kolkata Knight Riders,bat,normal,0,Kolkata Knight Riders,41,0,JH Kallis,Sheikh Zayed Stadium,M Erasmus,RK Illingworth,NaN
460,461,2014,Abu Dhabi,2014-04-18,Sunrisers Hyderabad,Rajasthan Royals,Rajasthan Royals,field,normal,0,Rajasthan Royals,0,4,AM Rahane,Sheikh Zayed Stadium,BF Bowden,RK Illingworth,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
462,463,2014,NaN,2014-04-19,Kolkata Knight Riders,Delhi Daredevils,Kolkata Knight Riders,bat,normal,0,Delhi Daredevils,0,4,JP Duminy,Dubai International Cricket Stadium,Aleem Dar,VA Kulkarni,NaN
474,475,2014,NaN,2014-04-28,Royal Challengers Bangalore,Kings XI Punjab,Kings XI Punjab,field,normal,0,Kings XI Punjab,0,5,Sandeep Sharma,Dubai International Cricket Stadium,BF Bowden,S Ravi,NaN
461,462,2014,NaN,2014-04-19,Mumbai Indians,Royal Challengers Bangalore,Royal Challengers Bangalore,field,normal,0,Royal Challengers Bangalore,0,7,PA Patel,Dubai International Cricket Stadium,Aleem Dar,AK Chaudhary,NaN
468,469,2014,NaN,2014-04-25,Sunrisers Hyderabad,Delhi Daredevils,Sunrisers Hyderabad,bat,normal,0,Sunrisers Hyderabad,4,0,AJ Finch,Dubai International Cricket Stadium,M Erasmus,S Ravi,NaN


##  Step 27: Sorting by City and Date

**What is happening here?**
Sorts the DataFrame first by `city` (A to Z), then by `date` (oldest to newest) within each city. This creates a chronological ordering of matches within each location.

**Real-world use:**
Sorting by geography and then time is a very common pattern in event data analysis — store transactions by location then by date, server logs by region then by timestamp, sports events by venue then by date.

**Purpose in this notebook:**
Produces a clean chronological history of matches for each city — useful for studying how each venue's usage evolved over time.

In [ ]:
data.sort_values(['city','date'])

,id,season,city,date,team1,team2,toss_winner,toss_decision,result,dl_applied,winner,win_by_runs,win_by_wickets,player_of_match,venue,umpire1,umpire2,umpire3
457,458,2014,Abu Dhabi,2014-04-16,Kolkata Knight Riders,Mumbai Indians,Kolkata Knight Riders,bat,normal,0,Kolkata Knight Riders,41,0,JH Kallis,Sheikh Zayed Stadium,M Erasmus,RK Illingworth,NaN
459,460,2014,Abu Dhabi,2014-04-18,Chennai Super Kings,Kings XI Punjab,Chennai Super Kings,bat,normal,0,Kings XI Punjab,0,6,GJ Maxwell,Sheikh Zayed Stadium,RK Illingworth,C Shamshuddin,NaN
460,461,2014,Abu Dhabi,2014-04-18,Sunrisers Hyderabad,Rajasthan Royals,Rajasthan Royals,field,normal,0,Rajasthan Royals,0,4,AM Rahane,Sheikh Zayed Stadium,BF Bowden,RK Illingworth,NaN
464,465,2014,Abu Dhabi,2014-04-21,Chennai Super Kings,Delhi Daredevils,Chennai Super Kings,bat,normal,0,Chennai Super Kings,93,0,SK Raina,Sheikh Zayed Stadium,RK Illingworth,C Shamshuddin,NaN
470,471,2014,Abu Dhabi,2014-04-26,Royal Challengers Bangalore,Rajasthan Royals,Rajasthan Royals,field,normal,0,Rajasthan Royals,0,6,PV Tambe,Sheikh Zayed Stadium,HDPK Dharmasena,C Shamshuddin,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
466,467,2014,NaN,2014-04-23,Chennai Super Kings,Rajasthan Royals,Rajasthan Royals,field,normal,0,Chennai Super Kings,7,0,RA Jadeja,Dubai International Cricket Stadium,HDPK Dharmasena,RK Illingworth,NaN
468,469,2014,NaN,2014-04-25,Sunrisers Hyderabad,Delhi Daredevils,Sunrisers Hyderabad,bat,normal,0,Sunrisers Hyderabad,4,0,AJ Finch,Dubai International Cricket Stadium,M Erasmus,S Ravi,NaN
469,470,2014,NaN,2014-04-25,Mumbai Indians,Chennai Super Kings,Mumbai Indians,bat,normal,0,Chennai Super Kings,0,7,MM Sharma,Dubai International Cricket Stadium,BF Bowden,M Erasmus,NaN
474,475,2014,NaN,2014-04-28,Royal Challengers Bangalore,Kings XI Punjab,Kings XI Punjab,field,normal,0,Kings XI Punjab,0,5,Sandeep Sharma,Dubai International Cricket Stadium,BF Bowden,S Ravi,NaN


##  Step 28: Per-Column Sort Direction with `ascending` as a List

**What is happening here?**
When `sort_values()` receives a list of columns, you can pass a corresponding list to `ascending` to specify the sort direction **independently for each column**. Here `ascending=[True, False]` would sort the first column ascending and second descending (though in this example both columns are the same, so the result is just ascending).

**Real-world use:**
Mixed-direction sorting is very useful in real reporting — for example, sort by department (A to Z) but by salary (highest first) within each department. This is equivalent to `ORDER BY dept ASC, salary DESC` in SQL.

**Purpose in this notebook:**
Demonstrates the syntax for specifying different sort orders for different columns in a multi-column sort.

In [ ]:
data.sort_values(['city','city'],ascending=[True,False])

,id,season,city,date,team1,team2,toss_winner,toss_decision,result,dl_applied,winner,win_by_runs,win_by_wickets,player_of_match,venue,umpire1,umpire2,umpire3
457,458,2014,Abu Dhabi,2014-04-16,Kolkata Knight Riders,Mumbai Indians,Kolkata Knight Riders,bat,normal,0,Kolkata Knight Riders,41,0,JH Kallis,Sheikh Zayed Stadium,M Erasmus,RK Illingworth,NaN
459,460,2014,Abu Dhabi,2014-04-18,Chennai Super Kings,Kings XI Punjab,Chennai Super Kings,bat,normal,0,Kings XI Punjab,0,6,GJ Maxwell,Sheikh Zayed Stadium,RK Illingworth,C Shamshuddin,NaN
460,461,2014,Abu Dhabi,2014-04-18,Sunrisers Hyderabad,Rajasthan Royals,Rajasthan Royals,field,normal,0,Rajasthan Royals,0,4,AM Rahane,Sheikh Zayed Stadium,BF Bowden,RK Illingworth,NaN
464,465,2014,Abu Dhabi,2014-04-21,Chennai Super Kings,Delhi Daredevils,Chennai Super Kings,bat,normal,0,Chennai Super Kings,93,0,SK Raina,Sheikh Zayed Stadium,RK Illingworth,C Shamshuddin,NaN
470,471,2014,Abu Dhabi,2014-04-26,Royal Challengers Bangalore,Rajasthan Royals,Rajasthan Royals,field,normal,0,Rajasthan Royals,0,6,PV Tambe,Sheikh Zayed Stadium,HDPK Dharmasena,C Shamshuddin,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
466,467,2014,NaN,2014-04-23,Chennai Super Kings,Rajasthan Royals,Rajasthan Royals,field,normal,0,Chennai Super Kings,7,0,RA Jadeja,Dubai International Cricket Stadium,HDPK Dharmasena,RK Illingworth,NaN
468,469,2014,NaN,2014-04-25,Sunrisers Hyderabad,Delhi Daredevils,Sunrisers Hyderabad,bat,normal,0,Sunrisers Hyderabad,4,0,AJ Finch,Dubai International Cricket Stadium,M Erasmus,S Ravi,NaN
469,470,2014,NaN,2014-04-25,Mumbai Indians,Chennai Super Kings,Mumbai Indians,bat,normal,0,Chennai Super Kings,0,7,MM Sharma,Dubai International Cricket Stadium,BF Bowden,M Erasmus,NaN
474,475,2014,NaN,2014-04-28,Royal Challengers Bangalore,Kings XI Punjab,Kings XI Punjab,field,normal,0,Kings XI Punjab,0,5,Sandeep Sharma,Dubai International Cricket Stadium,BF Bowden,S Ravi,NaN


##  Step 29: Removing Duplicates with `drop_duplicates()`

**What is happening here?**
`drop_duplicates(subset=['city'])` keeps only the **first occurrence** of each unique city, dropping all subsequent rows with the same city value. This reduces the 636-row DataFrame to just 31 rows — one per unique city.

**Real-world use:**
`drop_duplicates()` is essential for data cleaning. It removes redundant records from datasets, useful when extracting unique values — a list of unique customers, unique product categories, or in this case, a list of all unique cities that hosted IPL matches.

**Purpose in this notebook:**
Produces a list of the 31 distinct cities where IPL matches have been played, using the first match in each city as the representative row.

In [ ]:
data.drop_duplicates(subset=['city'])

,id,season,city,date,team1,team2,toss_winner,toss_decision,result,dl_applied,winner,win_by_runs,win_by_wickets,player_of_match,venue,umpire1,umpire2,umpire3
0,1,2017,Hyderabad,2017-04-05,Sunrisers Hyderabad,Royal Challengers Bangalore,Royal Challengers Bangalore,field,normal,0,Sunrisers Hyderabad,35,0,Yuvraj Singh,"Rajiv Gandhi International Stadium, Uppal",AY Dandekar,NJ Llong,NaN
1,2,2017,Pune,2017-04-06,Mumbai Indians,Rising Pune Supergiant,Rising Pune Supergiant,field,normal,0,Rising Pune Supergiant,0,7,SPD Smith,Maharashtra Cricket Association Stadium,A Nand Kishore,S Ravi,NaN
2,3,2017,Rajkot,2017-04-07,Gujarat Lions,Kolkata Knight Riders,Kolkata Knight Riders,field,normal,0,Kolkata Knight Riders,0,10,CA Lynn,Saurashtra Cricket Association Stadium,Nitin Menon,CK Nandan,NaN
3,4,2017,Indore,2017-04-08,Rising Pune Supergiant,Kings XI Punjab,Kings XI Punjab,field,normal,0,Kings XI Punjab,0,6,GJ Maxwell,Holkar Cricket Stadium,AK Chaudhary,C Shamshuddin,NaN
4,5,2017,Bangalore,2017-04-08,Royal Challengers Bangalore,Delhi Daredevils,Royal Challengers Bangalore,bat,normal,0,Royal Challengers Bangalore,15,0,KM Jadhav,M Chinnaswamy Stadium,NaN,NaN,NaN
6,7,2017,Mumbai,2017-04-09,Kolkata Knight Riders,Mumbai Indians,Mumbai Indians,field,normal,0,Mumbai Indians,0,4,N Rana,Wankhede Stadium,Nitin Menon,CK Nandan,NaN
10,11,2017,Kolkata,2017-04-13,Kings XI Punjab,Kolkata Knight Riders,Kolkata Knight Riders,field,normal,0,Kolkata Knight Riders,0,8,SP Narine,Eden Gardens,A Deshmukh,NJ Llong,NaN
14,15,2017,Delhi,2017-04-15,Delhi Daredevils,Kings XI Punjab,Delhi Daredevils,bat,normal,0,Delhi Daredevils,51,0,CJ Anderson,Feroz Shah Kotla,YC Barde,Nitin Menon,NaN
31,32,2017,Chandigarh,2017-04-28,Sunrisers Hyderabad,Kings XI Punjab,Kings XI Punjab,field,normal,0,Sunrisers Hyderabad,26,0,Rashid Khan,"Punjab Cricket Association IS Bindra Stadium, ...",Nitin Menon,CK Nandan,NaN
48,49,2017,Kanpur,2017-05-10,Gujarat Lions,Delhi Daredevils,Delhi Daredevils,field,normal,0,Delhi Daredevils,0,2,SS Iyer,Green Park,YC Barde,AK Chaudhary,NaN


##  Step 30: Deduplication with Multiple Columns

**What is happening here?**
`drop_duplicates(subset=['city','winner'])` treats a combination of `city` and `winner` as the unique identifier — it keeps only the first match where a particular team won in a particular city. The result has 198 rows.

**Real-world use:**
Multi-column deduplication is used when uniqueness is defined by a combination of fields — e.g., a customer can only have one record per product category, or an employee can have only one record per project.

**Purpose in this notebook:**
Reveals that across all cities and teams, there are 198 unique (city, winner) combinations — indicating which teams have won at least once in each city.

In [ ]:
data.drop_duplicates(subset=['city','winner']).shape

(198, 18)

##  Step 31: Finding the IPL Champion of Each Season

**What is happening here?**
This is a clever use of `drop_duplicates()` with `keep='last'`. The IPL dataset is ordered chronologically, so the **last match of each season** is the final. By keeping only the last occurrence of each season and selecting `season` and `winner`, we extract the **champion of each IPL season** from 2008 to 2017.

**Real-world use:**
Using `keep='last'` with `drop_duplicates()` is a useful trick for extracting the most recent record per group — e.g., the last transaction per customer, the most recent status per order, or the final score per game.

**Purpose in this notebook:**
Produces a clean table showing all 10 IPL champions: Rajasthan Royals (2008), Deccan Chargers (2009), Chennai Super Kings (2010 & 2011), Kolkata Knight Riders (2012 & 2014), Mumbai Indians (2013, 2015, 2017), and Sunrisers Hyderabad (2016).

In [ ]:
data.drop_duplicates('season',keep='last')[['season','winner']].sort_values('season')

,season,winner
116,2008,Rajasthan Royals
173,2009,Deccan Chargers
233,2010,Chennai Super Kings
306,2011,Chennai Super Kings
380,2012,Kolkata Knight Riders
456,2013,Mumbai Indians
516,2014,Kolkata Knight Riders
575,2015,Mumbai Indians
635,2016,Sunrisers Hyderabad
58,2017,Mumbai Indians


##  Step 32: Grouping and Counting with `groupby()` and `size()`

**What is happening here?**
`groupby('city')` splits the DataFrame into groups — one group per unique city. `.size()` counts the number of rows in each group. The result is then sorted in descending order to show which cities hosted the most matches.

**Real-world use:**
`groupby()` is the Pandas equivalent of SQL's `GROUP BY`. It is one of the most powerful tools for aggregating data — counting transactions per store, calculating total sales per product category, or finding average salaries per department. It is used in virtually every data analysis pipeline.

**Purpose in this notebook:**
Reveals that **Mumbai** hosted the most IPL matches (85), followed by Bangalore (66) and Kolkata (61). This reflects historical venue usage patterns in the IPL.

In [ ]:
data.groupby('city').size().sort_values(ascending=False)

,0
city,
Mumbai,85
Bangalore,66
Kolkata,61
Delhi,60
Hyderabad,49
Chennai,48
Chandigarh,46
Jaipur,33
Pune,32


##  Step 33: Grouped Aggregation — Counting Non-Null Values

**What is happening here?**
`data.groupby('city')['winner'].count()` groups by city and then counts only the **non-null** values in the `winner` column for each city. This is slightly different from `.size()`, which counts all rows including nulls. The difference is small here (since `winner` has only 3 null values), but the distinction matters in real data with significant missing values.

**Real-world use:**
Using `.count()` after `groupby()` ensures you're counting actual data, not missing entries. In datasets with many NaN values, `count()` and `size()` can give very different results.

**Purpose in this notebook:**
Confirms match counts per city based on valid (non-null) winner records, showing Mumbai still leads with 85 valid results.

In [ ]:
data.groupby('city')['winner'].count().sort_values(ascending=False)

,winner
city,
Mumbai,85
Bangalore,64
Kolkata,61
Delhi,59
Hyderabad,49
Chennai,48
Chandigarh,46
Jaipur,33
Pune,32


##  Step 34: Renaming Columns with `rename()`

**What is happening here?**
`rename(columns={'city': 'place'})` renames the column `city` to `place`. The `columns` parameter takes a dictionary mapping old names to new names. Note that without `inplace=True`, this returns a new DataFrame and the original `data` is unchanged.

**Real-world use:**
Renaming columns is a standard data cleaning step. Column names from raw data sources are often unclear, abbreviated, or follow different naming conventions. Renaming makes code more readable and ensures consistent naming across a project.

**Purpose in this notebook:**
Demonstrates how to rename a column. The display shows `place` instead of `city` but the underlying `data` variable is not modified.

In [ ]:
data.rename(columns={'city':'place'})

,id,season,place,date,team1,team2,toss_winner,toss_decision,result,dl_applied,winner,win_by_runs,win_by_wickets,player_of_match,venue,umpire1,umpire2,umpire3
0,1,2017,Hyderabad,2017-04-05,Sunrisers Hyderabad,Royal Challengers Bangalore,Royal Challengers Bangalore,field,normal,0,Sunrisers Hyderabad,35,0,Yuvraj Singh,"Rajiv Gandhi International Stadium, Uppal",AY Dandekar,NJ Llong,NaN
1,2,2017,Pune,2017-04-06,Mumbai Indians,Rising Pune Supergiant,Rising Pune Supergiant,field,normal,0,Rising Pune Supergiant,0,7,SPD Smith,Maharashtra Cricket Association Stadium,A Nand Kishore,S Ravi,NaN
2,3,2017,Rajkot,2017-04-07,Gujarat Lions,Kolkata Knight Riders,Kolkata Knight Riders,field,normal,0,Kolkata Knight Riders,0,10,CA Lynn,Saurashtra Cricket Association Stadium,Nitin Menon,CK Nandan,NaN
3,4,2017,Indore,2017-04-08,Rising Pune Supergiant,Kings XI Punjab,Kings XI Punjab,field,normal,0,Kings XI Punjab,0,6,GJ Maxwell,Holkar Cricket Stadium,AK Chaudhary,C Shamshuddin,NaN
4,5,2017,Bangalore,2017-04-08,Royal Challengers Bangalore,Delhi Daredevils,Royal Challengers Bangalore,bat,normal,0,Royal Challengers Bangalore,15,0,KM Jadhav,M Chinnaswamy Stadium,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
631,632,2016,Raipur,2016-05-22,Delhi Daredevils,Royal Challengers Bangalore,Royal Challengers Bangalore,field,normal,0,Royal Challengers Bangalore,0,6,V Kohli,Shaheed Veer Narayan Singh International Stadium,A Nand Kishore,BNJ Oxenford,NaN
632,633,2016,Bangalore,2016-05-24,Gujarat Lions,Royal Challengers Bangalore,Royal Challengers Bangalore,field,normal,0,Royal Challengers Bangalore,0,4,AB de Villiers,M Chinnaswamy Stadium,AK Chaudhary,HDPK Dharmasena,NaN
633,634,2016,Delhi,2016-05-25,Sunrisers Hyderabad,Kolkata Knight Riders,Kolkata Knight Riders,field,normal,0,Sunrisers Hyderabad,22,0,MC Henriques,Feroz Shah Kotla,M Erasmus,C Shamshuddin,NaN
634,635,2016,Delhi,2016-05-27,Gujarat Lions,Sunrisers Hyderabad,Sunrisers Hyderabad,field,normal,0,Sunrisers Hyderabad,0,4,DA Warner,Feroz Shah Kotla,M Erasmus,CK Nandan,NaN


##  Step 35: Setting a Custom Index with `set_index()`

**What is happening here?**
`set_index('city', inplace=True)` replaces the default integer index (0, 1, 2...) with the values from the `city` column. With `inplace=True`, the original DataFrame is modified directly. Now city names serve as the row labels.

**Real-world use:**
Setting a meaningful index improves data access speed and makes label-based lookups more intuitive. For example, if you set a customer ID as the index, you can retrieve all data for a specific customer instantly with `data.loc['C001']`. Time-series data often uses dates as the index.

**Purpose in this notebook:**
Sets city as the index, demonstrating how the DataFrame structure changes — `city` is no longer a regular column but becomes the row identifier.

In [ ]:
from re import I
data.set_index('city',inplace=True)

##  Step 36: Viewing the DataFrame After Setting a Custom Index

**What is happening here?**
Calling `data.head()` again shows the effect of the `set_index()` operation. The `city` column is now displayed as the row label on the left, and the remaining 17 columns are the data columns.

**Real-world use:**
It is always good practice to verify your operations by re-examining the DataFrame after making structural changes. This confirms that the index was set correctly.

**Purpose in this notebook:**
Visually confirms the new structure — city names (Hyderabad, Pune, Rajkot, etc.) now appear as the index instead of numbers 0, 1, 2...

In [ ]:
data.head()

,id,season,date,team1,team2,toss_winner,toss_decision,result,dl_applied,winner,win_by_runs,win_by_wickets,player_of_match,venue,umpire1,umpire2,umpire3
city,,,,,,,,,,,,,,,,,
Hyderabad,1,2017,2017-04-05,Sunrisers Hyderabad,Royal Challengers Bangalore,Royal Challengers Bangalore,field,normal,0,Sunrisers Hyderabad,35,0,Yuvraj Singh,"Rajiv Gandhi International Stadium, Uppal",AY Dandekar,NJ Llong,NaN
Pune,2,2017,2017-04-06,Mumbai Indians,Rising Pune Supergiant,Rising Pune Supergiant,field,normal,0,Rising Pune Supergiant,0,7,SPD Smith,Maharashtra Cricket Association Stadium,A Nand Kishore,S Ravi,NaN
Rajkot,3,2017,2017-04-07,Gujarat Lions,Kolkata Knight Riders,Kolkata Knight Riders,field,normal,0,Kolkata Knight Riders,0,10,CA Lynn,Saurashtra Cricket Association Stadium,Nitin Menon,CK Nandan,NaN
Indore,4,2017,2017-04-08,Rising Pune Supergiant,Kings XI Punjab,Kings XI Punjab,field,normal,0,Kings XI Punjab,0,6,GJ Maxwell,Holkar Cricket Stadium,AK Chaudhary,C Shamshuddin,NaN
Bangalore,5,2017,2017-04-08,Royal Challengers Bangalore,Delhi Daredevils,Royal Challengers Bangalore,bat,normal,0,Royal Challengers Bangalore,15,0,KM Jadhav,M Chinnaswamy Stadium,NaN,NaN,NaN


##  Step 37: Resetting the Index with `reset_index()`

**What is happening here?**
`reset_index('city', inplace=True)` reverses the `set_index()` operation — it moves the `city` index back to become a regular column, and restores the default integer index (0, 1, 2...).

**Real-world use:**
`reset_index()` is frequently needed after `groupby()` operations, after filtering, or after setting a custom index when you want to revert to a clean default integer index. It is also useful when you need to pass a DataFrame to a function or library that expects a standard integer index.

**Purpose in this notebook:**
Restores the original DataFrame structure with `city` as a column and numeric row indices.

In [ ]:
data.reset_index('city',inplace=True)

##  Step 38: Verifying the Restored Structure

**What is happening here?**
Calling `data.head()` after `reset_index()` confirms that the DataFrame is back to its original structure — `city` is a regular column again, and rows are indexed 0, 1, 2, 3, 4.

**Real-world use:**
Always verify data transformations by inspecting the result. Developing a habit of checking the DataFrame after structural operations helps catch mistakes early.

**Purpose in this notebook:**
Confirms successful restoration of the default structure after the `set_index()` / `reset_index()` demonstration.

In [ ]:
data.head()

,city,id,season,date,team1,team2,toss_winner,toss_decision,result,dl_applied,winner,win_by_runs,win_by_wickets,player_of_match,venue,umpire1,umpire2,umpire3
0,Hyderabad,1,2017,2017-04-05,Sunrisers Hyderabad,Royal Challengers Bangalore,Royal Challengers Bangalore,field,normal,0,Sunrisers Hyderabad,35,0,Yuvraj Singh,"Rajiv Gandhi International Stadium, Uppal",AY Dandekar,NJ Llong,NaN
1,Pune,2,2017,2017-04-06,Mumbai Indians,Rising Pune Supergiant,Rising Pune Supergiant,field,normal,0,Rising Pune Supergiant,0,7,SPD Smith,Maharashtra Cricket Association Stadium,A Nand Kishore,S Ravi,NaN
2,Rajkot,3,2017,2017-04-07,Gujarat Lions,Kolkata Knight Riders,Kolkata Knight Riders,field,normal,0,Kolkata Knight Riders,0,10,CA Lynn,Saurashtra Cricket Association Stadium,Nitin Menon,CK Nandan,NaN
3,Indore,4,2017,2017-04-08,Rising Pune Supergiant,Kings XI Punjab,Kings XI Punjab,field,normal,0,Kings XI Punjab,0,6,GJ Maxwell,Holkar Cricket Stadium,AK Chaudhary,C Shamshuddin,NaN
4,Bangalore,5,2017,2017-04-08,Royal Challengers Bangalore,Delhi Daredevils,Royal Challengers Bangalore,bat,normal,0,Royal Challengers Bangalore,15,0,KM Jadhav,M Chinnaswamy Stadium,NaN,NaN,NaN


##  Step 39: Converting a Series to a DataFrame with `reset_index()`

**What is happening here?**
`value_counts()` returns a Series where the team names are the index. Calling `.reset_index()` on this Series converts it to a proper 2-column DataFrame with columns `winner` and `count`, which is much easier to work with for further analysis or visualization.

**Real-world use:**
Converting Series results to DataFrames is a common pattern when you need the result as a table — for example, to join it with another DataFrame, export it to CSV, or pass it to a visualization library like Seaborn or Plotly.

**Purpose in this notebook:**
Produces a clean two-column table of team names and their win counts, ready for visualization or export.

In [ ]:
data['winner'].value_counts().reset_index()

,winner,count
0,Mumbai Indians,92
1,Chennai Super Kings,79
2,Kolkata Knight Riders,77
3,Royal Challengers Bangalore,73
4,Kings XI Punjab,70
5,Rajasthan Royals,63
6,Delhi Daredevils,62
7,Sunrisers Hyderabad,42
8,Deccan Chargers,29
9,Gujarat Lions,13


##  Step 40: Confirming Dataset Size

**What is happening here?**
Calling `data.shape` again at this point in the notebook confirms that the DataFrame still has 636 rows and 18 columns. None of the operations since Step 6 have permanently altered the number of rows or columns in `data`.

**Real-world use:**
Re-checking `shape` throughout an analysis is a good defensive habit. It helps confirm that operations like merges, filters, or deduplication had the expected impact on dataset size.

**Purpose in this notebook:**
A quick sanity check to confirm data integrity after all the transformations performed above.

In [ ]:
data.shape

(636, 18)

##  Step 41: Dropping Rows Where ALL Values Are Missing

**What is happening here?**
`dropna(how='all')` removes rows where **every single column** is `NaN`. With `how='all'`, a row is only dropped if there is no data at all in that row. The result is still 636 rows — meaning there are no completely empty rows in this dataset.

**Real-world use:**
`dropna(how='all')` is a safe first cleanup step. Completely empty rows can appear from bad data exports or manual data entry errors. This operation removes the most obvious garbage without discarding rows that have only some missing values.

**Purpose in this notebook:**
Confirms there are no fully empty rows in the IPL dataset.

In [ ]:
data.dropna(how='all').shape

(636, 18)

##  Step 42: Dropping Rows with ANY Missing Value

**What is happening here?**
`dropna(how='any')` drops **every row that has at least one NaN** anywhere. The result is 0 rows — meaning there is no single row in this entire dataset where all 18 columns have complete data (since `umpire3` is entirely null).

**Real-world use:**
`dropna(how='any')` is used when you need a fully complete dataset for algorithms that cannot handle missing values (e.g., many machine learning models). However, as seen here, it can be too aggressive — removing all data is rarely useful. It is best used after you have already handled the major missing columns.

**Purpose in this notebook:**
Demonstrates that the `umpire3` column (which is 100% null) makes `how='any'` eliminate every row. This highlights the importance of understanding which columns have missing data before using this option.

In [ ]:
data.dropna(how='any').shape

(0, 18)

##  Step 43: Dropping Columns with Missing Values

**What is happening here?**
`dropna(axis=1)` removes **columns** (not rows) that contain any missing values. Setting `axis=1` switches the operation from rows to columns. The result shows 12 columns remain out of 18 — meaning 6 columns had at least one missing value.

**Real-world use:**
Dropping columns with missing values is useful when you need clean features for machine learning models, or when a column has too many missing values to be useful. However, you should always consider whether the column is important before dropping it.

**Purpose in this notebook:**
Reveals that 6 columns have at least some missing data (`city`, `winner`, `player_of_match`, `umpire1`, `umpire2`, `umpire3`), reducing the dataset to 12 complete columns.

In [ ]:
data.dropna(axis=1).shape

(636, 12)

##  Step 44: Dropping Rows with Missing Values (axis=0)

**What is happening here?**
`dropna(axis=0)` is the default behavior — it drops **rows** that contain any `NaN`. This is identical to `dropna(how='any')` which we already saw returns 0 rows. `axis=0` is the default and refers to rows.

**Real-world use:**
Understanding the `axis` parameter is fundamental in Pandas. `axis=0` operates on rows, `axis=1` on columns. This applies to many other Pandas operations like `mean()`, `sum()`, `apply()`, etc.

**Purpose in this notebook:**
Reinforces the concept that `axis=0` (default) means "operate along rows" — in this case, dropping rows.

In [ ]:
data.dropna(axis=0).shape

(0, 18)

## Step 45: Targeted `dropna()` with `subset`

**What is happening here?**
`dropna(subset=['umpire3'])` drops rows only where the `umpire3` column is null. Since `umpire3` is 100% null (all 636 values are missing), every row is dropped and the result is 0 rows.

**Real-world use:**
The `subset` parameter is very useful when you only care about missingness in specific columns. For example, `dropna(subset=['customer_id'])` would only remove records where the customer identifier is missing, leaving other incomplete records intact.

**Purpose in this notebook:**
Confirms that `umpire3` is completely empty — all 636 rows are dropped when we require this column to be non-null.

In [ ]:
data.dropna(subset=['umpire3']).shape

(0, 18)

##  Step 46: Targeted `dropna()` on the `winner` Column

**What is happening here?**
`dropna(subset=['winner'])` drops only the rows where `winner` is null. The result is 633 rows — confirming that 3 matches in the dataset have no recorded winner (likely due to ties or abandoned matches).

**Real-world use:**
Filtering out records with missing target variables is a critical step before any analysis focused on outcomes. If you're analyzing win patterns, you should work only with matches that have a declared winner.

**Purpose in this notebook:**
Useful before any winner-focused analysis to ensure we are only working with complete, valid match results.

In [ ]:
data.dropna(subset=['winner']).shape

(633, 18)

## Step 47: Filling Missing Values with `fillna()`

**What is happening here?**
`fillna(0)` replaces **all** `NaN` values in the entire DataFrame with the number `0`. Note that this returns a new DataFrame — the original `data` is unchanged. In the output, `umpire3` now shows `0.0` and missing umpire names show `0` as well.

**Real-world use:**
`fillna()` is a key data imputation technique. Common strategies include filling with `0` (for numeric counts), filling with the mean/median of the column, forward-filling from the previous value (`method='ffill'`), or filling with a specific placeholder string. The right strategy depends on the context.

**Purpose in this notebook:**
Demonstrates the simplest form of missing value imputation — replacing all NaNs with zero. This is appropriate for numeric columns but less appropriate for text columns.

In [ ]:
data.fillna(0)

,city,id,season,date,team1,team2,toss_winner,toss_decision,result,dl_applied,winner,win_by_runs,win_by_wickets,player_of_match,venue,umpire1,umpire2,umpire3
0,Hyderabad,1,2017,2017-04-05,Sunrisers Hyderabad,Royal Challengers Bangalore,Royal Challengers Bangalore,field,normal,0,Sunrisers Hyderabad,35,0,Yuvraj Singh,"Rajiv Gandhi International Stadium, Uppal",AY Dandekar,NJ Llong,0.0
1,Pune,2,2017,2017-04-06,Mumbai Indians,Rising Pune Supergiant,Rising Pune Supergiant,field,normal,0,Rising Pune Supergiant,0,7,SPD Smith,Maharashtra Cricket Association Stadium,A Nand Kishore,S Ravi,0.0
2,Rajkot,3,2017,2017-04-07,Gujarat Lions,Kolkata Knight Riders,Kolkata Knight Riders,field,normal,0,Kolkata Knight Riders,0,10,CA Lynn,Saurashtra Cricket Association Stadium,Nitin Menon,CK Nandan,0.0
3,Indore,4,2017,2017-04-08,Rising Pune Supergiant,Kings XI Punjab,Kings XI Punjab,field,normal,0,Kings XI Punjab,0,6,GJ Maxwell,Holkar Cricket Stadium,AK Chaudhary,C Shamshuddin,0.0
4,Bangalore,5,2017,2017-04-08,Royal Challengers Bangalore,Delhi Daredevils,Royal Challengers Bangalore,bat,normal,0,Royal Challengers Bangalore,15,0,KM Jadhav,M Chinnaswamy Stadium,0,0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
631,Raipur,632,2016,2016-05-22,Delhi Daredevils,Royal Challengers Bangalore,Royal Challengers Bangalore,field,normal,0,Royal Challengers Bangalore,0,6,V Kohli,Shaheed Veer Narayan Singh International Stadium,A Nand Kishore,BNJ Oxenford,0.0
632,Bangalore,633,2016,2016-05-24,Gujarat Lions,Royal Challengers Bangalore,Royal Challengers Bangalore,field,normal,0,Royal Challengers Bangalore,0,4,AB de Villiers,M Chinnaswamy Stadium,AK Chaudhary,HDPK Dharmasena,0.0
633,Delhi,634,2016,2016-05-25,Sunrisers Hyderabad,Kolkata Knight Riders,Kolkata Knight Riders,field,normal,0,Sunrisers Hyderabad,22,0,MC Henriques,Feroz Shah Kotla,M Erasmus,C Shamshuddin,0.0
634,Delhi,635,2016,2016-05-27,Gujarat Lions,Sunrisers Hyderabad,Sunrisers Hyderabad,field,normal,0,Sunrisers Hyderabad,0,4,DA Warner,Feroz Shah Kotla,M Erasmus,CK Nandan,0.0


##  Step 48: Inspecting the `umpire3` Column Before Filling

**What is happening here?**
Displaying `data['umpire3']` shows that every value in this column is `NaN` (Not a Number / missing). The dtype is `float64` — Pandas uses float to represent integer columns that contain NaN values (since Python integers can't hold NaN natively).

**Real-world use:**
Before deciding how to handle missing values in a specific column, you should inspect it to understand the extent and pattern of missingness. A completely empty column like this might signal that it should simply be dropped rather than filled.

**Purpose in this notebook:**
Sets up context for the next step by confirming that `umpire3` has no real data.

In [ ]:
data['umpire3']

,umpire3
0,NaN
1,NaN
2,NaN
3,NaN
4,NaN
...,...
631,NaN
632,NaN
633,NaN
634,NaN


##  Step 49: Column-Specific `fillna()`

**What is happening here?**
`data['umpire3'].fillna(0)` fills the missing values in just the `umpire3` column with `0`, replacing all `NaN` with `0.0`. This is applied only to this single column rather than the whole DataFrame.

**Real-world use:**
Column-specific filling is preferred over whole-DataFrame filling because different columns often need different imputation strategies. For example, numeric columns might be filled with the column mean, while categorical columns might be filled with the mode or a placeholder like `'Unknown'`.

**Purpose in this notebook:**
Shows how to target a specific column for `fillna()` — a much more controlled and realistic approach than filling the entire DataFrame at once.

In [ ]:
data['umpire3'].fillna(0)

,umpire3
0,0.0
1,0.0
2,0.0
3,0.0
4,0.0
...,...
631,0.0
632,0.0
633,0.0
634,0.0


##  Step 50: Filling with a String Value and `inplace=True`

**What is happening here?**
`data['umpire3'].fillna("yes", inplace=True)` fills all NaN values in `umpire3` with the string `"yes"` and **modifies the DataFrame in place** (no new copy is created). A warning is shown by Pandas about incompatible dtypes (float64 vs string), which is expected behavior.

**Real-world use:**
Using `inplace=True` applies the change directly to the existing DataFrame without needing to reassign it. The `inplace` parameter appears throughout Pandas (e.g., `sort_values`, `rename`, `set_index`). In modern best practices, creating a new variable is often preferred over `inplace=True` for clarity and to avoid chained assignment warnings.

**Purpose in this notebook:**
Demonstrates both string filling and the `inplace` modification pattern. After this operation, every row in `umpire3` shows `"yes"` as confirmed in the next cell.

In [ ]:
data['umpire3'].fillna("yes",inplace=True)

/tmp/ipykernel_16026/3460394837.py:1: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'yes' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  data['umpire3'].fillna("yes",inplace=True)


##  Step 51: Final Verification

**What is happening here?**
The final `data.head()` confirms the result of all operations: `city` is restored as a regular column, all index operations have been undone, and `umpire3` now shows `"yes"` for every row instead of `NaN`.

**Real-world use:**
Always do a final check after your data cleaning and transformation pipeline to verify the DataFrame is in the expected state before saving it or passing it to the next stage (visualization, modeling, reporting).

**Purpose in this notebook:**
Serves as the final confirmation that all data manipulation steps were applied correctly, wrapping up the data cleaning workflow.

In [ ]:
data.head()

,city,id,season,date,team1,team2,toss_winner,toss_decision,result,dl_applied,winner,win_by_runs,win_by_wickets,player_of_match,venue,umpire1,umpire2,umpire3
0,Hyderabad,1,2017,2017-04-05,Sunrisers Hyderabad,Royal Challengers Bangalore,Royal Challengers Bangalore,field,normal,0,Sunrisers Hyderabad,35,0,Yuvraj Singh,"Rajiv Gandhi International Stadium, Uppal",AY Dandekar,NJ Llong,yes
1,Pune,2,2017,2017-04-06,Mumbai Indians,Rising Pune Supergiant,Rising Pune Supergiant,field,normal,0,Rising Pune Supergiant,0,7,SPD Smith,Maharashtra Cricket Association Stadium,A Nand Kishore,S Ravi,yes
2,Rajkot,3,2017,2017-04-07,Gujarat Lions,Kolkata Knight Riders,Kolkata Knight Riders,field,normal,0,Kolkata Knight Riders,0,10,CA Lynn,Saurashtra Cricket Association Stadium,Nitin Menon,CK Nandan,yes
3,Indore,4,2017,2017-04-08,Rising Pune Supergiant,Kings XI Punjab,Kings XI Punjab,field,normal,0,Kings XI Punjab,0,6,GJ Maxwell,Holkar Cricket Stadium,AK Chaudhary,C Shamshuddin,yes
4,Bangalore,5,2017,2017-04-08,Royal Challengers Bangalore,Delhi Daredevils,Royal Challengers Bangalore,bat,normal,0,Royal Challengers Bangalore,15,0,KM Jadhav,M Chinnaswamy Stadium,NaN,NaN,yes
